### ISSUE TITLE
[RFC] Speed up Torch AMP and cliip grad norm operations by flattening all the tensors behind the scene

### ISSUE THREAD

**Issue by blefaudeux** [Lightning-AI/lightning#11053]
## 🚀 Feature
- Flatten all the tensors for a given model (per parameter group for instance) in the background (not user visible)
- Replace the per parameter tensors by tensor views
- Speed up:
- - optimizer step (per parameter group, in most cases it will become just a few matmuls)
- - torch AMP grad scaler (one kernel on each way instead of one per parameter)
- - clip grad norm

edit: 
Depending on the targeted speed up, the above would need to apply to the parameters, the grads, or both

### Motivation
Per parameter gradient scaling (same scale for all) and grad norm take a very significant time, benchmarked to 15-20% on a single GPU with [this small test](https://github.com/facebookresearch/xformers/blob/main/examples/microGPT.py#L314). Doing the above would be the equivalent to fusing multiple kernel calls into one, which is provably faster

### Pitch
- make this an option for lightning (this will break if the user changes tensors manually during training, which is probably not a good idea anyway, but there are valid corner cases)
- use a [flat param wrapper](https://github.com/facebookresearch/fairscale/blob/main/fairscale/nn/misc/flatten_params_wrapper.py#L39). This is done in several places in [Fairscale](https://github.com/facebookresearch/fairscale/), if that helps
- do the same with the grads (**this is definitely a limitation of the proposal, open for other suggestions, my memory is a bit fuzzy around that**. This nukes the grad=None option) 
- derive a custom grad scaler from pytorch's which works on the flat buffer instead of working on all the parameters
- change the anomaly detection to work on the flat buffer, instead of going through all the grads

### Alternatives
Not doing the above, that's fine but slow

cc @borda @akihironitta @SeanNaren @tchaton

**Comment by SeanNaren** [comment]
I really like this! I think we can have an opt in approach for this optimization, maybe passed to the Trainer. cc @ananthsub

**Comment by tchaton** [comment]
Hey @blefaudeux, this is a great idea. I wondered if you would be interested in making a contribution ?

**Comment by blefaudeux** [comment]
sure, not a lot of time at the moment but I could have a look and give it a shot !

**Comment by rishabbala** [comment]
Heu @tchaton I can take this up if it is still open/required.

**Comment by Borda** [comment]
> Heu @tchaton I can take this up if it is still open/required.

Sure! It will be great =)

**Comment by rishabbala** [comment]
Thanks @Borda . Could you give a quick overview of where I should start digging and how this could be implemented?

**Comment by awaelchli** [comment]
Should we close this old issue? With parameter flattening now in PyTorch FSDP, I think we should be covered. Or do you see any benefit of having a separete implementation for other use cases?
